buildbot/osuosl/master/master.cfg

```diff
@@ -26,11 +26,18 @@ c['buildbotNetUsageData'] = None
 import config
 reload(config)
 
+# Detect if the BUILDMASTER_TEST environment variable is set and non-zero.
+# This will be used to enable a local testing mode.
+buildmaster_test = os.environ.get('BUILDMASTER_TEST')
+test_mode = bool(buildmaster_test) and buildmaster_test != '0'
+config.options['Internal'] = {}
+config.options['Internal']['test_mode'] = str(test_mode)
+
 ####### Workers
 
 c['workers'] = config.workers.get_all()
 
-c['protocols'] = {'pb': {'port': 9990}}
+c['protocols'] = {'pb': {'port': "tcp:9990:interface=127.0.0.1" if test_mode else 9990}}
```

Comment:
I not 100% sure on this but I believe putting interface=127.0.0.1 will make test buildmaster inaccessible from the outside world which is not always the case. For example, if we are running this test master in a cloud vps and want our out of network workers to have access to it.

Reply:
That's correct, my thinking is to default to something that's as secure as possible. I use ssh port forwarding in this case, but I think `socat` could also work if you just want to quickly expose the port to the outside world with no auth needed. Or of course editing the command line yourself.

I can document ssh port forwarding and/or socat for this and the web interface in the followon patch to https://llvm.org/docs/HowToAddABuilder.html that describes testing mode.

Perhaps I'm being over-cautious? But on the other hand:
* This mode uses a default password for connections, meaning it's easy to connect to the buildmaster and thus exploitation is much more plausible vs a production buildmaster where attackers shouldn't be able to connect
* People often run a mode like this on developer machines or within secure networks, so the potential impact is high vs a production buildmaster where (hopefully!) time and thought have gone into appropriate permissions / sandboxing / firewalling.
* For most cases where you're interactively testing a buildmaster and set of builders, I'd imagine using ssh port forwarding does the job fine.

Anyway, that's my thinking - other suggestions welcome!